### ISSUE TITLE
bug: Security risks associated with using innerHTML

### ISSUE THREAD

**Issue by cjorasch** [ionic-team/ionic#18065]
# Bug Report

**Ionic version:**
[x] **4.2.0**

**Current behavior:**
Ionic core uses `innerHTML` in a number of places that result in security risks when used in conjunction with user entered content.

From: https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML

> Warning: If your project is one that will undergo any form of security review, using innerHTML most likely will result in your code being rejected. For example, if you use innerHTML in a browser extension and submit the extension to addons.mozilla.org, it will not pass the automated review process.

This was originally submitted as issue #14586 but that issue was closed.

Several Ionic components have properties that are string values but are handled as HTML content when rendering the component.

An example usage where there is a security risk is:

```tsx
    const alert = await this.alertController.create({
      message: 'Hello ' + this.name,
    });
```

If the user is able to enter the value for `name` then they are able to insert raw HTML into the dom.

Specific components where this issue exists include:
- `<ion-alert>.message`
- `<ion-searchbar>.placeholder`
- `<ion-infinite-scroll-content>.loadingText`
- `<ion-refresher-content>.pullingText`
- `<ion-refresher-content>.refershingText`

**Other issues:**

In addition to security risks, it is also difficult for the developer to know that the property is handled as raw html.  For example, the `ion-alert` message property is defined as `The main message to be displayed in the alert` in the documentation.  If you do read through the source code you will likely not know if is raw html.

This makes it less likely the developer will know to escape any user entered values which makes it likely that there will be security holes.

It also can cause display issues when used with user entered data.  For example:

```tsx
    const alert = await this.alertController.create({
      message: 'Hello ' + this.name,
    });
```

If `name` includes `<` or other html characters then the message will not render correctly.

**Proposed solution:**

The ability to render content beyond a simple string is very valuable so there should be an ability to specify formatted content.

A potential solution is to use properties that can contain JSX values.  For example:

```tsx
@Prop() message: JSX.Element;
```

This allows the value to be a simple string value to avoid complicating the interface.  It also allows the value to be more complex content when appropriate.  For example:

```tsx
<my-component message={<span>Hello <b>user</b></span>} />
```

This technique could also be used for other properties that could benefit from increased formatting flexibility.  For example: alert.header, toast.message, etc.

**Comment by liamdebeasi** [comment]
Hi there,

Thanks for the issue. I am able to reproduce using the following:

```
const alert = await this.alertCtrl.create({
  message: 'Hello <img src="x" onerror="alert(1)">'
});
    
await alert.present();
```

(Source: https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#Security_considerations)

I am going to discuss with the team the best way to address this issue.

Thanks!

**Comment by cjorasch** [comment]
I noticed this is being worked on and glad to see that the security risk will be closed.

I am curious, however, why a sanitization approach is being used rather than simply leveraging JSX.

In particular, there are a few downsides with the sanitization approach:
- The whitelist of elements and attributes will be one more thing to learn and lead to ongoing developer requests to extend the whitelist.
- Special string properties that can contain some html are not obvious from the component interface where the property appears to be a normal string.
- The developer still needs to escape user content to handle reserved characters (<, &, etc.) for it to display correctly.
- Putting content in a `<slot>` vs. specifying as a property uses a different technique.

If the property is declared as `@Prop() message: JSX.Element;`
- Leverages existing developer experience with JSX.
- Any html content can be represented
- Security is maintained
- User content is escaped
- the component interface clearly defines what is valid
- Simple string values can be used as is
- No extra code required

Is there some downside to the JSX approach?

**Comment by liamdebeasi** [comment]
Hi there,

Thanks for the follow up. The main issue with the JSX approach is it does not stop bad actors from executing arbitrary JavaScript (the main thing we want to avoid), while still allowing safe HTML to be included.

The following would still be allowed:

```
<a href="javascript:alert('JavaScript Executed!')">A Harmless Link</a>
```

 I agree that the HTML/string usage could be much clearer in our documentation, but ultimately we do want to allow users to include small bits of HTML in some of these props (i.e. showing an image in an `ion-alert`).

Sanitize             |  JSX.Element via `innerHTML` | JSX.Element via rendering inside of a `div`
:--------------:|:-------------:|:-------------:
![image](https://user-images.githubusercontent.com/2721089/56591967-bd3fb400-65b7-11e9-9124-8822331c9fa9.png)  |  ![image](https://user-images.githubusercontent.com/2721089/56592076-ee1fe900-65b7-11e9-9e49-46356710f3e0.png) | ![image](https://user-images.githubusercontent.com/2721089/56594010-3c35ec00-65ba-11e9-84e4-076283a6d3c9.png)


I am using the following as input for the examples above:
```
<a href="javascript:alert(`oh no!`);">click me!</a> <img src="http://lorempixel.com/400/200/cats/"> 123 < 150 1 & 2 & 3 Lorem ipsum dolor sit amet
```

Notice in the `JSX.Element via innerHTML` image, the link is still clickable. Clicking the link would cause JavaScript to execute. Additionally, rendering the message via `<div>{this.message</div>` will cause all HTML to be forced to string which is not something we want either. In the `Sanitize` image, the JavaScript is removed from the link, and the rest of the valid HTML is still rendered. 

In general, we provide sanitization methods for our components that use `innerHTML`, but it is ultimately up to the user to ensure that they do not introduce _new_ XSS vulnerabilities when they build with Ionic.

We will also be creating a security guide in our docs that explains these changes for other developers.

Please let me know if you have any questions.
Thanks!

**Comment by cjorasch** [comment]
When supporting JSX you would not use innerHTML because, as you mentioned, that would still have a security risk.  But there is no need to use innerHTML at all.

```tsx
@Prop() message: JSX.Element;

render() {
  return <span>Hello {this.message}!</span>;
}
```

This is what happens everywhere else you are rendering user content.  This approach allows the developer to apply markup where appropriate without needing to escape user content.

I think the primary time that the sanitization approach is useful is when you want the end user to be able to enter limited html content (similar to markdown) but this seems like a less likely use case.  If for some reason it is desired it could be handled with a component that wraps partial html and then used anywhere that JSX is supported.  If I enter a value into a `<input>` or load a value from a database generally the desire is to render it the way it was entered rather than to assume it is html.

If I enter `John Smith <john@smith.com>` into a text field I generally expect it to be rendered the same way in both displayed content and alerts.

**Comment by liamdebeasi** [comment]
Hi there,

With components like `ion-alert` we want the user to be able to enter limited HTML content, so JSX.Element is not particularly useful in this case as `this.message` would be converted to a string before being rendered.

**Comment by cjorasch** [comment]
When would the user enter limited HTML content and how would that relate to displaying an alert vs. any other rendering?

For example, in an app like the ionic conference application.  If users build the schedule of events then each event has a title.  The title is displayed in lots of other components so any html characters will be displayed as is.  If the title is `This -> that` then it will appear that way on the pages.  If the user is able to delete an event the app might display a message "Are you sure you want to delete This -> that?".  If the alert assumes html then it gets rendered differently.  If you think the value should be able to contain html then it would need to be rendered differently everywhere else.

It seems like the more likely use case is:

```tsx
message: <span>Are you sure you want to delete <b>{event.title}</b>?</span>
```

In this case the alert can display formatted content correctly but still apply the same formatting to use content that is used everywhere else.

**Comment by liamdebeasi** [comment]
With the sanitize approach, your content is still rendered the same. All we do differently is we strip out things like `javascript:` or `onclick` handlers. Supplying characters such as `<` or `&` still render properly, so providing `This -> that` to the `message` prop would still work correctly.

Additionally, we can’t export as JSX.Element because we don’t require JSX to use Ionic core. Ionic core can be used inside of an Angular application and needs to be able to be passed in as such.

After discussing with the team, we are going to move forward with this sanitize approach.

Thanks for bringing this issue to our attention! I will post here when the updates have been merged in.

**Comment by simonhaenisch** [comment]
Not really security related, but I agree that having the possibility of passing in JSX would be very useful here... in fact that is something I already tried, in the same way that @cjorasch explained above in https://github.com/ionic-team/ionic/issues/18065#issuecomment-485877477, i. e., the user entering the HTML/JSX would be the developer working with Ionic (where having JSX available is not unlikely), not the end user of the app.

Would it be possible to type the prop as e. g. 

```tsx
@Prop() message: string | JSX.Element;
```

or does that also make it not work in a non-jsx setup? Passing HTML as a string when working with JSX feels so wrong 🤓 Maybe it's possible to conditionally avoid using `innerHTML` if JSX is passed in? But not sure how it would be possible to distinguish between string and jsx type in the renderer at runtime.

**Comment by liamdebeasi** [comment]
Hi everyone,

I think my previous comment was a little bit vague, so let me attempt to clarify 🙂  

We aren't able to implement the JSX approach for two reasons:
1. Switching from an only string input to an only JSX.Element input would be considered a breaking change, and we only do those for major releases (4.0, 5.0, 6.0, etc)
2. Switching from an only string input to an only JSX.Element input would prevent users who do not use JSX from using certain Ionic components (For example, a person using Ionic components in an vanilla JS application).

Supporting both string and JSX.Element is something worth discussing with the team (in terms of feasibility and technical limitations), so I will certainly do that. However, adding JSX.Element would be considered a feature and is out of scope for this security fix.

Our focus is on ensuring that existing component inputs are sanitized where necessary.

For anyone who would like to see JSX.Element supported in components like `ion-alert`, I recommend creating a [Feature Request](https://github.com/ionic-team/ionic/issues/new?assignees=&labels=&template=feature_request.md&title=feat%3A+) as this issue will likely be closed once the security fix is in place.

Thanks!

**Comment by liamdebeasi** [comment]
Hi there,

This has been fixed via https://github.com/ionic-team/ionic/pull/18083. We will be providing patch releases for all minor versions of Ionic 4 later today (4.0.x, 4.1.x, etc).

Thanks for bringing this to our attention, and thanks for using Ionic!
